### ISSUE TITLE
kubernetes is CPU-hungry on minikube

### ISSUE THREAD

**Issue by rcorre** [kubernetes/kubernetes#48948]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:

When running kubernetes 1.7 on minikube, CPU usage for the `VBoxHeadless` process is constantly around 100%.

**What you expected to happen**:

CPU usage closer to 10%, as it is when running 1.6.4

**How to reproduce it (as minimally and precisely as possible)**:

```
minikube start --kubernetes-version v1.7.0
top
minikube stop
minikube start --kubernetes-version v1.6.4
top
```

**Anything else we need to know?**:

This could be a minikube bug, but since CPU usage changes drastically between kubernetes versions (but the same minikube version), I figured it might be a kubernetes issue.

**Environment**:

- Kubernetes version (use `kubectl version`): 1.7.0 vs 1.6.4
- Cloud provider or hardware configuration**: local
- OS (e.g. from /etc/os-release): ArchLinux
- Kernel (e.g. `uname -a`): 4.9.33-1-lts 
- Install tools: huh?
- Others: minikube version: v0.20.0

**Comment by rcorre** [comment]
/sig scalability

Really no idea if that is the right group, but none seemed completely appropriate

**Comment by timbunce** [comment]
See kubernetes/minikube#1158.

**Comment by timbunce** [comment]
This is still a problem with the minikube v0.28.0.
For example, on my macOS 10.13.5, VBoxHeadless uses ~50% cpu with nothing running.
That's still far about the 10% as it was when running 1.6.4.

Could someone change the title to remove the reference to 1.7?

/remove-lifecycle stale

**Comment by GregSilverman** [comment]
Same issue with v0.26.0.

**Comment by yeluolei** [comment]
same issue in mac with 0.28

**Comment by michilu** [comment]
Same issue with v0.28.1.

**Comment by ypresto** [comment]
Maybe related; issue about Kubernetes and CPU of docker-for-mac:
https://github.com/docker/for-mac/issues/2601

**Comment by briandealwis** [comment]
I see this high CPU (~30% at idle) with MacOS both with Docker for Mac's bundled Kubernetes and with minikube.

@lizrice [described an approach for installing k8s using Vagrant](https://medium.com/@lizrice/kubernetes-in-vagrant-with-kubeadm-21979ded6c63) without minikube, but still with VirtualBox.  She noted that she was still seeing high CPU usage. minikube may not be the issue here.

**Comment by briandealwis** [comment]
I tried using minikube with the new `docker-machine-driver-vmware` driver (https://github.com/kubernetes/minikube/pull/2606) to start in VMWare Fusion, to see if this usage was something tied to VirtualBox or hyper kit, and it has the same ~30% CPU usage, though with a bit more jitter.  I tried with both Kubernetes v1.10.0 and v1.11.2.

**Comment by Laski** [comment]
/remove-lifecycle stale

This is still happening to me in v0.30.0

**Comment by timbunce** [comment]
@rcorre, or someone with permission, please edit the title to remove the reference to 1.7.

**Comment by Laski** [comment]
Is there any information I can give to help debugging this?

**Comment by erulabs** [comment]
Can someone from sig/scalability take a look at this? I would love to help out - can test, reproduce, explore, report... This is becoming quite a blocker for local Kubernetes development!

**Comment by leahciMic** [comment]
Wondering what the best way to develop locally for kubernetes? The base load and slowness of minikube is a deal breaker for rapid development especially when using a laptop. 

I have thought about:

1. Running it on a server dedicated to this. This could be a potential solution to run it on a server on my network, but it has it's own issues like being more difficult to update docker images etc, or developing from a different network, (for example, coffee shop).

2. Running services on my machine (elasticsearch, mysql, redis, etc) and using them in develop, and then using containerized services in production. I have already discounted this idea, as I would like to have my develop environment match my production environment.

3. Run with vm-driver=none

But all of these issues have there own problems. 

I assume that with the popularity of kubernetes and minikube that there must be a way to run a local cluster without the level of resource consumption that many people are seeing.

Is there an alternative to minikube that has lower overheads?

Basically an idle mysql server on my laptop is no big deal, and I don't notice it. An idling minikube cluster (without any services running mind you), is definitely felt.

**Comment by dims** [comment]
@leahciMic did you try kind? https://github.com/kubernetes-sigs/kind

**Comment by almike** [comment]
I got 30-40% constant usage with minikube, 10-20% with microk8s.
I am new to kube, not sure where to start, but I could dig some logs to help.
would be nice to see this solved as we need to run ingress locally on mac and the only way to do so is to have kube running.

**Comment by StefPac** [comment]
I am experiencing the same problem with minikube v1.5.2 on Darwin 10.14.6
and Kubernetes v1.16.2 on Docker '18.09.9' using Oracle VM VirtualBox Headless Interface 6.0.14

**Comment by ankurkanoria** [comment]
Similar issue, vastly hinders local development agility. Setup a new barebones single node development cluster using kubeadm (v1.17) inside a reasonably beefy vm 4 cores 4gb ram (with kvm tuning like raw disk virtio). Idle state cpu usage is around 30%, spikes to 50-60% during container creation/deletion. Controller manager and scheduler often timeout waiting for Apiserver and go to crashloopbackoff state (40+ restarts after 10 hours of ~idling). Similar issues across a range of hardware configurations on different Linux distributions.

While searching around, came across quite a few stackoverflow posts and github issues complaining of similar observations. While I understand that kube is optimized for running at scale, it does - non-trivially - hurt local development.

What will be great if there are some resources pointing us towards configuration changes that reduce resource consumption at the expense of reducing cluster reliability, e.g. health check/poll frequencies, etc. As developers working on local machines, many of us might be happy to take that tradeoff.

**Comment by nyrahul** [comment]
Same issue (100% cpu for vboxheadless)
* minikube v1.6.1 on Ubuntu 18.04 arch x86_64
* Linux Kernel 5.3.0
* K8s v1.17.0
* vboxheadless v6.1.2

Attached is the perf report (recorded with few secs while vboxheadless is at 100% CPU).
[vboxheadless-perf-report.txt](https://github.com/kubernetes/kubernetes/files/4210468/vboxheadless-perf-report.txt)

**Comment by sunshine69** [comment]
Hm seeing this thread is a bit painfull. People keep posting, reporting but seems no one to care. Issues keep coming stale.

I probably run away from the whole kubeXXX all together for such behaviour.

Oh well maybe a attitude - submit a patch or shut up?

<shrug and go away>

**Comment by sunshine69** [comment]
And if someone ask me why I have nothing to do but posting here, then you can see below. But looking above makes me even not bothering to say anything ....

Wake up people. If you do so, you lose users, your project will be less and less popular and eventually phased out ......

```
top - 09:48:30 up 1 day, 12:29,  1 user,  load average: 0.38, 0.79, 0.94
Tasks: 340 total,   1 running, 339 sleeping,   0 stopped,   0 zombie
%Cpu(s):  3.2 us,  1.2 sy,  0.0 ni, 94.6 id,  0.8 wa,  0.0 hi,  0.1 si,  0.0 st
MiB Mem :  15891.9 total,   1112.2 free,   6238.8 used,   8540.9 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.   8825.4 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                   
 3333 libvirt+  20   0 8822744   2.7g  20904 S     26.2  17.3   3:25.87 qemu-system-x86                           
 3474 stevek    20   0 1202132  98308  52932 S   2.3   0.6   9:46.95 Xorg                                      
 4153 stevek    20   0  975464  53264  37600 S   2.0   0.3   0:07.13 gnome-terminal-                           
32747 stevek    20   0 9003620 244924  90080 S   1.3   1.5   1:20.64 chrome                                    
 3748 stevek    20   0 3863612 283392  98096 S   1.0   1.7  14:22.97 gnome-shell                               
 5367 stevek    20   0 1756812 528440 314792 S   0.7   3.2  26:00.70 chrome                         

```

**Comment by tstromberg** [comment]
FWIW, minikube has vastly improved CPU overhead since v1.8 - a reduction of 19% in 2020 alone.

On a modern developer machine like a MacBook Pro, minikube with hyperkit now consumes roughly 6% of available system CPU (30-40% of a single CPU core): https://docs.google.com/spreadsheets/d/1qzgVsZ9y0zqCjoQlN_LGJH3MUMqrVPexzNhdB2jzBqU/edit#gid=1614668143 

It isn't perfect, but I feel like this issue can be closed, or at least moved to the minikube repo. In 2020, we'll be focusing our eyes on reducing usage in apiserver and etcd, which is where most of minikube's CPU cycles are now spent.

**Comment by ravihugo** [comment]
@tstromberg That is great news. I hope the apiserver/etcd usage can be lowered!

**Comment by oomichi** [comment]
As https://github.com/kubernetes/kubernetes/issues/48948#issuecomment-625049243 it seems fine to close this issue at this time.
In addition, this issue is for minikube. It is better to open an issue on https://github.com/kubernetes/minikube/issues to get more attention from minikube experts if necessary to re-open again.

/close

**Comment by Pictor13** [comment]
> It isn't perfect, but I feel like this issue can be closed, or at least moved to the minikube repo. In 2020, we'll be focusing our eyes on reducing usage in apiserver and etcd, which is where most of minikube's CPU cycles are now spent.

Thank you for your job guys!
Sorry for commenting on a closed issue @tstromberg , but I was wondering if there's a specific Kubernetes issue on the tracker that we can follow to see updates regarding these improvements to *apiserver* and *etcd*.
